File: app/src/main/java/com/example/bookworm/core/login/LoginActivity.java
package com.example.bookworm.core.login;

import android.content.Context;
import android.content.Intent;

import android.os.Bundle;

import android.util.Log;
import android.view.View;
import android.widget.ImageButton;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;

import com.example.bookworm.appLaunch.modules.MainViewModel;
import com.example.bookworm.bottomMenu.bookworm.BookWorm;
import com.example.bookworm.appLaunch.views.MainActivity;
import com.example.bookworm.R;
import com.example.bookworm.core.userdata.UserInfo;
import com.example.bookworm.core.internet.FBModule;
import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.FirebaseAuth;
import com.kakao.auth.AuthType;
import com.kakao.auth.Session;

public class LoginActivity extends AppCompatActivity {

    protected SessionCallback sessionCallback = new SessionCallback();
    private static final String TAG = "MainActivity";
    public static Context mContext;
    private FirebaseAuth mAuth;
    private FBModule fbModule;
    protected GoogleSignInAccount gsa;
    UserInfo userInfo;
    BookWorm bookworm;
    public static GoogleSignInClient gsi;
    Boolean isLogined = Boolean.FALSE;
    private int RC_SIGN_IN = 123;
    private MainViewModel mv;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        setContentView(R.layout.activity_login);
        super.onCreate(savedInstanceState);

        mContext = this;
        fbModule = new FBModule(mContext);
        isLogined = Boolean.FALSE; //카카오 이중로그인 방지
        mAuth = FirebaseAuth.getInstance();
        mv = new ViewModelProvider(this, new MainViewModel.Factory(this)).get(MainViewModel.class);
        startLogin();
    }

    private void startLogin() {
        Session session = Session.getCurrentSession();
        session.addCallback(sessionCallback);

        // 구글

        // Configure Google Sign In
        GoogleSignInOptions gso = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestIdToken(getString(R.string.server_client_id))
                .requestEmail()
                .build();
        gsi = GoogleSignIn.getClient(this, gso);
        gsa = GoogleSignIn.getLastSignedInAccount(this);
        // 구글 자동 로그인
        if (gsa != null) {
            signInGoogle();
            return;
        }
        ImageButton google_login_button = (ImageButton) findViewById(R.id.btn_login_google);
        google_login_button.setOnClickListener(it -> {
                    signInGoogle();
                }
        );


        // 카카오
        // 이미 로그인되어있고 세션이 살아있는 경우에만 작동
        if (Session.getCurrentSession().checkAndImplicitOpen() && isLogined == Boolean.TRUE) {
            Log.d(TAG, "onClick: 로그인 세션살아있음");
            // 카카오 자동 로그인
            sessionCallback.requestMe();
            return;
        }

        ImageButton kakao_login_button = (ImageButton) findViewById(R.id.btn_login_kakao);
        kakao_login_button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                // 카카오 로그인 시도 (창이 뜬다.)
                if (!Session.getCurrentSession().checkAndImplicitOpen()) {
                    session.open(AuthType.KAKAO_LOGIN_ALL, LoginActivity.this);
                    isLogined = Boolean.TRUE;
                }
            }
        });
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        // 세션 콜백 삭제
        Session.getCurrentSession().removeCallback(sessionCallback);
    }


    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        // 카카오톡|스토리 간편로그인 실행 결과를 받아서 SDK로 전달
        if (Session.getCurrentSession().handleActivityResult(requestCode, resultCode, data)) return;

        //구글 로그인시도시
        if (requestCode == RC_SIGN_IN) {
            Task<GoogleSignInAccount> task = GoogleSignIn.getSignedInAccountFromIntent(data);
            task.addOnCompleteListener(task1 -> {
                //회원의 정보를 가져옴
                GoogleSignInAccount account = task1.getResult();
                //회원가입 여부를 확인.
                userInfo = new UserInfo();
                userInfo.add(account);
                userInfo.setToken(account.getId());
                signUp(userInfo);
            });
        }
    }


    //메인액티비티로 이동
    public void move() {
        Intent intent = new Intent(this, MainActivity.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);
        startActivity(intent);
        this.finish();
    }


    //회원가입 함수
    public void signUp(UserInfo userInfo) {
        mv.getUser(userInfo.getToken(), true); //회원 여부 확인을 위한 회원정보 조회
        mv.getUserInfo().observe(this, it -> {
            if (null != userInfo.getUsername()) {
                //회원인 경우
                if (it != null) move();
                    //회원이 아닌 경우
                else {
                    //사용자 생성
                    mv.createUser(userInfo);
                    //액티비티 이동
                    mv.getData().observe(this, et -> {
                        if (et) move();
                    });
                }

            } else {
                Log.d("function signup", "nono token ");
            }
        });

    }

    // 구글 로그인 메소드
    protected void signInGoogle() {
        Intent signInIntent = gsi.getSignInIntent();
        startActivityForResult(signInIntent, RC_SIGN_IN);
    }
}

File: app/src/main/java/com/example/bookworm/core/login/LoginActivity.kt
package com.example.bookworm.core.login

import android.content.Intent
import android.os.Bundle
import android.util.Log
import androidx.activity.result.ActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModelProvider
import com.example.bookworm.R
import com.example.bookworm.appLaunch.views.MainActivity
import com.example.bookworm.bottomMenu.profile.UserInfoViewModel
import com.example.bookworm.bottomMenu.profile.views.PreferGenreActivity
import com.example.bookworm.core.userdata.UserInfo
import com.example.bookworm.databinding.ActivityLoginBinding
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInAccount
import com.google.android.gms.auth.api.signin.GoogleSignInClient
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.tasks.Task
import com.google.firebase.auth.AuthResult
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.GoogleAuthProvider
import com.google.firebase.messaging.FirebaseMessaging
import com.kakao.sdk.auth.model.OAuthToken
import com.kakao.sdk.common.model.AuthErrorCause
import com.kakao.sdk.common.model.ClientError
import com.kakao.sdk.common.model.ClientErrorCause
import com.kakao.sdk.user.UserApiClient


class LoginActivity : AppCompatActivity() {
    private val binding by lazy {
        ActivityLoginBinding.inflate(layoutInflater)
    }
    private var mAuth: FirebaseAuth? = null
    private var gsa: GoogleSignInAccount? = null
    var userInfo: UserInfo? = null
    var isLogined: Boolean = java.lang.Boolean.FALSE
    private val userViewModel by lazy {
        ViewModelProvider(
                this, UserInfoViewModel.Factory(this)
        )[UserInfoViewModel::class.java]
    }
    var startActivityResult = registerForActivityResult<Intent, ActivityResult>(
            ActivityResultContracts.StartActivityForResult()
    ) { result: ActivityResult ->
        if (result.resultCode == RESULT_OK) {
            val task = GoogleSignIn.getSignedInAccountFromIntent(result.data)
            task.addOnCompleteListener { task1: Task<GoogleSignInAccount> ->
                //회원의 정보를 가져옴
                val account = task1.result
                gsa = task1.result
                Log.d("구글 로그인 데이터", task1.result.idToken!!)
                //회원 정보 객체를 생성
                userInfo = UserInfo(
                        username = account.displayName!!,
                        token = account.id!!,
                        profileimg = try {
                            account.photoUrl.toString()!!
                        } catch (e: NullPointerException) {
                            ""
                        },
                        platform = "Google"
                )
                //회원 가입 함수로 데이터를 전달
                signUp(userInfo!!)
            }
        }
    }

    //카카오톡으로 로그인 할 수 없어 카카오 계정으로 로그인 할 경우 이용
    val callback: (OAuthToken?, Throwable?) -> Unit = { token, error ->
        if (error != null) {
            when (error.toString()) {
                AuthErrorCause.AccessDenied.toString() -> {
                    //접근이 거부됨 (동의 취소)
                }
                else -> {
                    //기타 에러
                }
            }


            //카카오계정으로 로그인 실패
        } else if (token != null) {
            //카카오계정으로 로그인 성공
            createUserByKakao()
        }
    }


    override fun onCreate(savedInstanceState: Bundle?) {
        setContentView(binding.root)
        super.onCreate(savedInstanceState)
        isLogined = java.lang.Boolean.FALSE //카카오 이중로그인 방지
        mAuth = FirebaseAuth.getInstance()
        startLogin()

    }

    private fun startLogin() {

        // 구글

        // Configure Google Sign In
        val gso = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestIdToken(getString(R.string.server_client_id))
                .requestEmail()
                .build()
        gsi = GoogleSignIn.getClient(this, gso)
        gsa = GoogleSignIn.getLastSignedInAccount(this)
        // 구글 자동 로그인
        if (gsa != null) {
            signInGoogle()
            return
        }
        binding.apply {
            btnLoginGoogle.setOnClickListener {
                signInGoogle()
            }
            btnLoginKakao.setOnClickListener {
                if (UserApiClient.instance.isKakaoTalkLoginAvailable(this@LoginActivity)) {
                    UserApiClient.instance.loginWithKakaoAccount(this@LoginActivity) { token, error ->
                        if (error != null) {

                            //사용자가 카카오톡 설치 후 디바이스 권한 요청 화면에서 로그인을 취소한 경우,
                            // 의도적인 로그인 취소로 보고 카카오 계저응로 로그인 시도 없이 로그인 취소로 처리
                            if (error is ClientError && error.reason == ClientErrorCause.Cancelled) {
                                return@loginWithKakaoAccount
                            }

                            //카카오톡에 연결된 카카오 계정이 없는 경우, 카카오 계정으로 로그인 시도
                            UserApiClient.instance.loginWithKakaoAccount(context = this@LoginActivity, callback = callback)
                        } else if (token != null) {
                            //카카오 톡으로 로그인 성공
                            createUserByKakao()
                        }
                    }
                } else UserApiClient.instance.loginWithKakaoAccount(context = this@LoginActivity, callback = callback)
            }

        }
    }


    private fun createUserByKakao() {
        UserApiClient.instance.me { user, error ->
            signUp(UserInfo(
                    username = user!!.kakaoAccount!!.profile!!.nickname!!,
                    token = user.id.toString(),
                    profileimg = user.kakaoAccount!!.profile!!.profileImageUrl!!,
                    platform = "Kakao"
            ))
        }
    }


    //메인액티비티로 이동
    private fun StartApplication(_iFlag: Int) {

        /** flag 0 == 가입, 1 == login */
        if(_iFlag == 0)
        {
            val intent = Intent(this, PreferGenreActivity::class.java)
            intent.putExtra("Login", 1)
            intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP
            startActivity(intent)
            finish()
        }
        if(_iFlag == 1)
        {
            val intent = Intent(this, MainActivity::class.java)
            startActivity(intent)
            finish()
//            intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP
        }
    }

    //회원가입 함수
    fun signUp(userInfo: UserInfo) {
        if (gsa != null) {
            val accessToken = gsa!!.idToken
            val credential = GoogleAuthProvider.getCredential(accessToken, null)
            mAuth!!.signInWithCredential(credential).addOnCompleteListener { task: Task<AuthResult> -> Log.d("로그인 완료", task.result.toString()) }
        }
        userViewModel.getUser(userInfo.token, true) //회원 여부 확인을 위한 회원정보 조회
        CheckFcm(userInfo)
        userViewModel.userInfoLiveData.observe(this) { userinfo: UserInfo? ->
            //회원인 경우
            if (userinfo!!.platform != null) StartApplication(1)
            else {
                val userInfoLiveData = MutableLiveData<Boolean>()
                //사용자 생성
                userViewModel.createUser(userInfo, userInfoLiveData)
                //액티비티 이동
                userInfoLiveData.observe(this) { et: Boolean -> if (et) StartApplication(0) }
            }
        }
    }

    private fun CheckFcm (userInfo: UserInfo) {
        userViewModel.getUser(userInfo!!.token, true)
        FirebaseMessaging.getInstance().token.addOnCompleteListener { task ->
            if (!task.isSuccessful) {
                if(userInfo!!.fCMtoken != task.result || userInfo!!.fCMtoken == null)
                    userInfo!!.fCMtoken = task.result
            }
        }
    }

    // 구글 로그인 메소드
    private fun signInGoogle() {
        val signInIntent = gsi!!.signInIntent
        startActivityResult.launch(signInIntent)
    }

    companion object {

        @JvmField
        var gsi: GoogleSignInClient? = null
    }
}